File: src/main/java/com/portofino/realtrainmodunofficial/block/LargeRailCoreBlock.java
package com.portofino.realtrainmodunofficial.block;

import com.portofino.realtrainmodunofficial.RealTrainModUnofficialComponents;
import com.portofino.realtrainmodunofficial.RealTrainModUnofficialBlockEntities;
import com.portofino.realtrainmodunofficial.RealTrainModUnofficialItems;
import com.portofino.realtrainmodunofficial.blockentity.LargeRailCoreBlockEntity;
import com.portofino.realtrainmodunofficial.blockentity.RailCollisionBlockEntity;
import com.portofino.realtrainmodunofficial.rail.util.RailMap;
import com.portofino.realtrainmodunofficial.rail.util.RailPosition;
import com.mojang.serialization.MapCodec;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import com.portofino.realtrainmodunofficial.item.WrenchItem;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.BaseEntityBlock;
import net.minecraft.world.level.block.RenderShape;
import net.minecraft.world.level.block.SoundType;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.Nullable;

public class LargeRailCoreBlock extends BaseEntityBlock {
    public static final MapCodec<LargeRailCoreBlock> CODEC = simpleCodec(LargeRailCoreBlock::new);
    private static final VoxelShape SHAPE = Shapes.box(0.0D, 0.0D, 0.0D, 1.0D, 0.0625D, 1.0D);

    public LargeRailCoreBlock(BlockBehaviour.Properties props) {
        super(props);
    }

    public LargeRailCoreBlock() {
        this(BlockBehaviour.Properties.of().sound(SoundType.METAL).strength(0.5F, 6.0F).noOcclusion());
    }

    @Override
    protected MapCodec<? extends BaseEntityBlock> codec() {
        return CODEC;
    }

    @Override
    public RenderShape getRenderShape(BlockState state) {
        return RenderShape.INVISIBLE;
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
        return SHAPE;
    }

    @Override
    public VoxelShape getCollisionShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
        return SHAPE;
    }

    @Override
    public VoxelShape getInteractionShape(BlockState state, BlockGetter level, BlockPos pos) {
        return SHAPE;
    }

    @Override
    public void onRemove(BlockState state, net.minecraft.world.level.Level level, BlockPos pos, BlockState newState, boolean isMoving) {
        if (!level.isClientSide && state.getBlock() != newState.getBlock()) {
            BlockEntity be = level.getBlockEntity(pos);
            if (be instanceof LargeRailCoreBlockEntity core) {
                RailMap[] maps = core.getAllRailMaps();
                if (maps.length > 0) {
                    // Prevent collision blocks from recursively trying to delete this core.
                    boolean prev = com.portofino.realtrainmodunofficial.rail.util.RailMap.suppressRailRemoval;
                    com.portofino.realtrainmodunofficial.rail.util.RailMap.suppressRailRemoval = true;
                    try {
                        for (RailMap map : maps) {
                            if (map != null) {
                                map.removeRailBlocks(level);
                            }
                        }
                        removeRemainingCollisionBlocks(level, pos, maps);
                    } finally {
                        com.portofino.realtrainmodunofficial.rail.util.RailMap.suppressRailRemoval = prev;
                    }
                }
            }
        }
        super.onRemove(state, level, pos, newState, isMoving);
    }

    @Nullable
    @Override
    public BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return new LargeRailCoreBlockEntity(pos, state);
    }

    public InteractionResult use(BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
        if (player.getItemInHand(hand).getItem() instanceof WrenchItem) {
            if (!level.isClientSide()) {
                if (level.getBlockEntity(pos) instanceof LargeRailCoreBlockEntity core) {
                    int total = core.getAllRailMaps().length;
                    if (total >= 2 && core.cycleSwitch()) {
                        player.displayClientMessage(
                            Component.literal("分岐切替: " + (core.getActiveSegmentIndex() + 1) + "/" + total + " 番線"),
                            true);
                    }
                }
            }
            return InteractionResult.sidedSuccess(level.isClientSide());
        }
        return InteractionResult.PASS;
    }

    @Override
    public ItemStack getCloneItemStack(LevelReader level, BlockPos pos, BlockState state) {
        // 中ボタン(ピックブロック): レールコアから種類+形状(長さ)をコピーした RailItem を返す。
        if (level.getBlockEntity(pos) instanceof LargeRailCoreBlockEntity core) {
            return createRailCloneStack(pos, core);
        }
        return ItemStack.EMPTY;
    }

    @Override
    public void neighborChanged(BlockState state, Level level, BlockPos pos, net.minecraft.world.level.block.Block block, BlockPos fromPos, boolean isMoving) {
        if (!level.isClientSide() && level.getBlockEntity(pos) instanceof LargeRailCoreBlockEntity core) {
            core.requestSwitchStateRefresh();
        }
        super.neighborChanged(state, level, pos, block, fromPos, isMoving);
    }

    @Override
    public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level level, BlockState state, BlockEntityType<T> type) {
        return createTickerHelper(type, RealTrainModUnofficialBlockEntities.LARGE_RAIL_CORE.get(), LargeRailCoreBlockEntity::tick);
    }

    private static void removeRemainingCollisionBlocks(Level level, BlockPos corePos, RailMap[] maps) {
        int minX = corePos.getX() - 2;
        int maxX = corePos.getX() + 2;
        int minY = corePos.getY() - 2;
        int maxY = corePos.getY() + 2;
        int minZ = corePos.getZ() - 2;
        int maxZ = corePos.getZ() + 2;
        for (RailMap map : maps) {
            if (map == null) {
                continue;
            }
            int split = RailMap.curveSplitForLength(map.getHorizontalPathLength());
            int samples = Math.max(16, split + 1);
            for (int i = 0; i < samples; i++) {
                int j = samples <= 1 ? 0 : (int) Math.round((double) split * i / (samples - 1));
                int index = Math.min(split, j);
                double[] point = map.getRailPos(split, index);
                int x = (int) Math.floor(point[1]);
                int y = (int) Math.floor(map.getRailHeight(split, index));
                int z = (int) Math.floor(point[0]);
                minX = Math.min(minX, x - 2);
                maxX = Math.max(maxX, x + 2);
                minY = Math.min(minY, y - 2);
                maxY = Math.max(maxY, y + 2);
                minZ = Math.min(minZ, z - 2);
                maxZ = Math.max(maxZ, z + 2);
            }
        }
        for (int x = minX; x <= maxX; x++) {
            for (int y = minY; y <= maxY; y++) {
                for (int z = minZ; z <= maxZ; z++) {
                    BlockPos scanPos = new BlockPos(x, y, z);
                    BlockEntity be = level.getBlockEntity(scanPos);
                    if (be instanceof RailCollisionBlockEntity collision && corePos.equals(collision.getCorePos())) {
                        level.removeBlock(scanPos, false);
                    }
                }
            }
        }
    }

    /**
     * Builds a copyable rail item from a placed rail core.
     * The returned stack reuses the same preview tag format as edited rails,
     * so right-clicking a marker can recreate the copied geometry.
     */
    public static ItemStack createRailCloneStack(BlockPos corePos, LargeRailCoreBlockEntity core) {
        ItemStack stack = new ItemStack(RealTrainModUnofficialItems.RAIL_ITEM.get());
        String railDefinitionId = core.getRailDefinitionId();
        if (railDefinitionId != null && !railDefinitionId.isBlank()) {
            stack.set(RealTrainModUnofficialComponents.SELECTED_MODEL_ID.get(), railDefinitionId);
        }

        RailPosition[] positions = core.getRailPositions();
        if (positions.length < 2 || positions[0] == null) {
            return stack;
        }

        CompoundTag preview = new CompoundTag();
        preview.putInt("X", corePos.getX());
        preview.putInt("Y", corePos.getY());
        preview.putInt("Z", corePos.getZ());
        // placeRailFromItem がコピー元の実ブロック不在でも復元できるように、
        // レンチ由来と同じ互換フラグを持たせる。
        preview.putBoolean("WrenchMode", true);
        preview.put("StartRP", positions[0].writeToNBT());

        ListTag segments = new ListTag();
        for (int i = 0; i + 1 < positions.length; i += 2) {
            RailPosition end = positions[i + 1];
            if (end == null) {
                continue;
            }
            CompoundTag segment = new CompoundTag();
            segment.put("EndRP", end.writeToNBT());
            segments.add(segment);
        }
        if (!segments.isEmpty()) {
            preview.putBoolean("BranchMode", segments.size() > 1);
            preview.put("RailSegments", segments);
            preview.put("EndRP", segments.getCompound(0).getCompound("EndRP"));
            stack.set(RealTrainModUnofficialComponents.RAIL_PREVIEW_START.get(), preview);
        }
        return stack;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/block/LargeRailCoreBlock.kt
package cc.mirukuneko.realtrainmodrenewed.block

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedComponents
import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities
import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedItems
import cc.mirukuneko.realtrainmodrenewed.blockentity.LargeRailCoreBlockEntity
import cc.mirukuneko.realtrainmodrenewed.blockentity.RailCollisionBlockEntity
import cc.mirukuneko.realtrainmodrenewed.compat.NbtCompat
import cc.mirukuneko.realtrainmodrenewed.rail.util.RailMap
import cc.mirukuneko.realtrainmodrenewed.rail.util.RailPosition
import cc.mirukuneko.realtrainmodrenewed.item.WrenchItem
import com.mojang.serialization.MapCodec
import net.minecraft.core.BlockPos
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.block.BaseEntityBlock
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.RenderShape
import net.minecraft.world.level.block.SoundType
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityTicker
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockBehaviour
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.shapes.CollisionContext
import net.minecraft.world.phys.shapes.Shapes
import net.minecraft.world.phys.shapes.VoxelShape
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

class LargeRailCoreBlock : BaseEntityBlock {
    companion object {
        val CODEC: MapCodec<LargeRailCoreBlock> = simpleCodec { LargeRailCoreBlock() }
        private val SHAPE = Shapes.box(0.0, 0.0, 0.0, 1.0, 0.0625, 1.0)

        @JvmStatic
        fun removeRailNetwork(level: Level, corePos: BlockPos?, core: LargeRailCoreBlockEntity?) {
            if (level == null || corePos == null) return
            val prev = RailMap.suppressRailRemoval.get()
            RailMap.suppressRailRemoval.set(true)
            try {
                val maps = core?.allRailMaps ?: emptyArray()
                for (map in maps) map?.removeRailBlocks(level)
                removeRemainingCollisionBlocks(level, corePos, maps)
                if (level.getBlockState(corePos).block is LargeRailCoreBlock) level.removeBlock(corePos, false)
            } finally { RailMap.suppressRailRemoval.set(prev) }
        }

        private fun removeRemainingCollisionBlocks(level: Level, corePos: BlockPos, maps: Array<RailMap>) {
            var minX = corePos.x - 2; var maxX = corePos.x + 2
            var minY = corePos.y - 2; var maxY = corePos.y + 2
            var minZ = corePos.z - 2; var maxZ = corePos.z + 2
            for (map in maps) {
                if (map == null) continue
                val split = RailMap.curveSplitForLength(map.getHorizontalPathLength())
                val samples = max(16, split + 1)
                for (i in 0 until samples) {
                    val j = if (samples <= 1) 0 else (split.toDouble() * i / (samples - 1)).roundToInt()
                    val idx = min(split, j)
                    val point = map.getRailPos(split, idx)
                    val x = floor(point[1]).toInt(); val y = floor(map.getRailHeight(split, idx)).toInt(); val z = floor(point[0]).toInt()
                    minX = min(minX, x - 2); maxX = max(maxX, x + 2)
                    minY = min(minY, y - 2); maxY = max(maxY, y + 2)
                    minZ = min(minZ, z - 2); maxZ = max(maxZ, z + 2)
                }
            }
            if (maps.isEmpty()) { minX = corePos.x - 128; maxX = corePos.x + 128; minY = corePos.y - 32; maxY = corePos.y + 32; minZ = corePos.z - 128; maxZ = corePos.z + 128 }
            for (x in minX..maxX) for (y in minY..maxY) for (z in minZ..maxZ) {
                val scanPos = BlockPos(x, y, z)
                val be = level.getBlockEntity(scanPos)
                if (be is RailCollisionBlockEntity && corePos == be.corePos) level.removeBlock(scanPos, false)
            }
        }

        @JvmStatic
        fun createRailCloneStack(corePos: BlockPos, core: LargeRailCoreBlockEntity): ItemStack {
            val stack = ItemStack(RealTrainModRenewedItems.RAIL_ITEM.get())
            val railDefId = core.railDefinitionId
            if (!railDefId.isNullOrBlank()) stack.set(RealTrainModRenewedComponents.SELECTED_MODEL_ID.get(), railDefId)
            val positions = core.railPositions ?: return stack
            if (positions.size < 2 || positions[0] == null) return stack
            val preview = CompoundTag()
            preview.putInt("X", corePos.x); preview.putInt("Y", corePos.y); preview.putInt("Z", corePos.z)
            preview.putBoolean("WrenchMode", true)
            preview.put("StartRP", positions[0]!!.writeToNBT())
            val segments = ListTag()
            var i = 0
            while (i + 1 < positions.size) {
                val start = positions[i] ?: run { i += 2; continue }
                val end = positions[i + 1] ?: run { i += 2; continue }
                val segment = CompoundTag()
                segment.put("StartRP", start.writeToNBT())
                segment.put("EndRP", end.writeToNBT())
                segments.add(segment)
                i += 2
            }
            if (!segments.isEmpty) {
                preview.putBoolean("BranchMode", segments.size > 1)
                preview.put("RailSegments", segments)
                preview.put("EndRP", NbtCompat.getCompound(NbtCompat.getCompound(segments, 0), "EndRP"))
                stack.set(RealTrainModRenewedComponents.RAIL_PREVIEW_START.get(), preview)
            }
            return stack
        }
    }

    constructor() : this(Properties.of().sound(SoundType.METAL).strength(0.5f, 6.0f).noOcclusion())
    constructor(props: BlockBehaviour.Properties) : super(props)

    override fun codec(): MapCodec<out BaseEntityBlock> = CODEC
    override fun getRenderShape(state: BlockState): RenderShape = RenderShape.INVISIBLE
    override fun getShape(state: BlockState, level: BlockGetter, pos: BlockPos, ctx: CollisionContext): VoxelShape = SHAPE
    override fun getCollisionShape(state: BlockState, level: BlockGetter, pos: BlockPos, ctx: CollisionContext): VoxelShape = SHAPE
    override fun getInteractionShape(state: BlockState, level: BlockGetter, pos: BlockPos): VoxelShape = SHAPE

    override fun affectNeighborsAfterRemoval(state: BlockState, level: ServerLevel, pos: BlockPos, isMoving: Boolean) {
        val be = level.getBlockEntity(pos)
        if (be is LargeRailCoreBlockEntity) removeRailNetwork(level, pos, be)
        super.affectNeighborsAfterRemoval(state, level, pos, isMoving)
    }

    override fun playerWillDestroy(level: Level, pos: BlockPos, state: BlockState, player: Player): BlockState {
        if (!level.isClientSide) { val be = level.getBlockEntity(pos); if (be is LargeRailCoreBlockEntity) removeRailNetwork(level, pos, be) }
        return super.playerWillDestroy(level, pos, state, player)
    }

    override fun newBlockEntity(pos: BlockPos, state: BlockState): BlockEntity = LargeRailCoreBlockEntity(pos, state)

    override fun useItemOn(stack: ItemStack, state: BlockState, level: Level, pos: BlockPos, player: Player, hand: InteractionHand, hit: BlockHitResult): InteractionResult {
        if (stack.item is WrenchItem && !level.isClientSide) {
            val core = level.getBlockEntity(pos) as? LargeRailCoreBlockEntity
            if (core != null && core.allRailMaps.size >= 2 && core.cycleSwitch())
                player.sendOverlayMessage(Component.literal("分岐切替: ${core.activeSegmentIndex + 1}/${core.allRailMaps.size} 番線"))
            return if (level.isClientSide) InteractionResult.SUCCESS else InteractionResult.SUCCESS_SERVER
        }
        return InteractionResult.PASS
    }

    override fun getCloneItemStack(level: LevelReader, pos: BlockPos, state: BlockState, includeData: Boolean): ItemStack {
        val be = level.getBlockEntity(pos)
        return if (be is LargeRailCoreBlockEntity) createRailCloneStack(pos, be) else ItemStack.EMPTY
    }

    override fun neighborChanged(state: BlockState, level: Level, pos: BlockPos, block: Block, orientation: net.minecraft.world.level.redstone.Orientation?, isMoving: Boolean) {
        if (!level.isClientSide) { val be = level.getBlockEntity(pos); if (be is LargeRailCoreBlockEntity) be.requestSwitchStateRefresh() }
        super.neighborChanged(state, level, pos, block, orientation, isMoving)
    }

    override fun <T : BlockEntity> getTicker(level: Level, state: BlockState, type: BlockEntityType<T>): BlockEntityTicker<T>? =
        null
}
